### ISSUE TITLE
ZFS broken on CentOS7

### ISSUE THREAD

**Issue by kevin-olbrich** [zfsonlinux/zfs#8140]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	|  CentOS
Distribution Version	|  7
Linux Kernel	|  3.10.0-862.14.4.el7
Architecture	| x86_64
ZFS Version	| 0.7.12
SPL Version	| 0.7.12

### Describe the problem you're observing
```
[root@example ~]# modprobe zfs
modprobe: ERROR: could not insert 'zfs': Invalid argument
```

### Describe how to reproduce the problem
Install zfs (dkms) or kmod-zfs on latest CentOS.

### Include any warning/errors/backtraces from the system logs

```
[ 1431.978577] zfs: disagrees about version of symbol vn_getattr
[ 1431.978587] zfs: Unknown symbol vn_getattr (err -22)
```

**Comment by loli10K** [comment]
> This seems to be only a problem with latest official CentOS 7 kernel.

FWIW the CentOS 7 builder running the same kernel version (3.10.0-862.14.4.el7) as yours seem to work just fine: http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/4752

**Comment by kevin-olbrich** [comment]
> > This seems to be only a problem with latest official CentOS 7 kernel.
> 
> FWIW the CentOS 7 builder running the same kernel version (3.10.0-862.14.4.el7) as yours seem to work just fine: http://build.zfsonlinux.org/builders/CentOS%207%20x86_64%20%28TEST%29/builds/4752

Are you sure? It lists 0.7.11 and the build ran before 0.7.12 was released.

**Comment by kpande** [comment]
please use [zfs-discuss](http://list.zfsonlinux.org) or IRC for support.

**Comment by cigoe** [comment]
@kevin-olbrich I just hit this as well and downgrading from 0.7.12 to 0.7.11 did not resolve the issue. Any chance you were able to find a solution? My current kernel version matches yours in the original issue.

**Comment by kevin-olbrich** [comment]
> please use [zfs-discuss](http://list.zfsonlinux.org) or IRC for support.

I don't need support, I just wanted to share that 0.7.12 is broken on CentOS. If I would need help, I would not have opened an issue on a repository.



> @kevin-olbrich I just hit this as well and downgrading from 0.7.12 to 0.7.11 did not resolve the issue. Any chance you were able to find a solution? My current kernel version matches yours in the original issue.

I had to find a solution asap, thats why I upgraded a machine from stock to elrepo kernel. I would not recommend this, as this is against the reason EL7 exists.

If you tried a downgrade and did not delete the kernel modules before, then dkms will most likely not overwrite them (just a guess). Maybe you need to clean up the modules/extra folder for your kernel.
Maybe you can just remove the kernel rpm with "no deps" and remove the module folder and then reinstall the kernel. This should not be a problem when the kernel is running.

Most of my ZFS upgrades failed, when the kernel received and update while ZFS did too. I then upgraded the kernel, reboot (without zfs support) and installed the new zfs packages. Seems like DKMS breaks when the current kernel does not match the most recent or when the kernel is installed after DKMS built a new module.

**Comment by artitj** [comment]
I ran into the same problem described by the original poster, though I am running the DKMS version. The source of the issue, I think, is that for some odd reason only some modules are getting rebuilt, so there were a mix of modules with 0.7.11 and 0.7.12. You can verify by running `modinfo znvpair` as an example.

I ended up solving the problem by first deleting all the modules in extra:
`rm /lib/modules/3.10.0-862.14.4.el7.x86_64/extra/*`

I wasn't able to rebuild the modules because dkms still thought they were installed, so I had to run these commands:
`dkms remove spl/0.7.12 -k 3.10.0-862.14.4.el7.x86_64`
`dkms remove zfs/0.7.12 -k 3.10.0-862.14.4.el7.x86_64`
I suppose that these 2 commands remove the need to do the rm command above, but I didn't test that. I'm going to try to skip the `rm` step above next time I run into this problem.

And then I rebuilt the modules:
`dkms install spl/0.7.12 -k 3.10.0-862.14.4.el7.x86_64`
`dkms install spl/0.7.12 -k 3.10.0-862.14.4.el7.x86_64`

At this point, you can verify that the module works now by trying to load it.

Previous to the above, I tried removing and then reinstalling all the zfs packages, but that didn't seem to work.

**Comment by cigoe** [comment]
I was able to resolve the above by [following the wiki](https://github.com/zfsonlinux/zfs/wiki/RHEL-and-CentOS) and switching to kABI-tracking kmod instead. I would assume the zfs module cleanup/reinstall would of resolved the issue but since we are stable and use the distribution kernel we're going to stay on kABI for now.

**Comment by kevin-olbrich** [comment]
Seems like you reproduced the exact same error. Indeed, it seems like my kernel was the same with 0.7.11 and I upgraded ZFS to 0.7.12.

Second, I had the same problem with DKMS. DKMS is made to compile modules against new kernels, which never worked for me (but does for acronis snapapi26).

Even after installing the elrepo kenel, dkms did not build the module even with "dkms autoinstall".



> I was able to resolve the above by [following the wiki](https://github.com/zfsonlinux/zfs/wiki/RHEL-and-CentOS) and switching to kABI-tracking kmod instead. I would assume the zfs module cleanup/reinstall would of resolved the issue but since we are stable and use the distribution kernel we're going to stay on kABI for now.

This did not work for me. Same error with the modules but I did not clean up modules/extra which might be the reason this did not work after the dkms-try.
